### ISSUE TITLE
Negative bignum OSSL_PARAM_BLD incorrect after retrieval

### ISSUE THREAD

**Issue by spoofedex** [openssl/openssl#17070]
Dear reader,

During some testing I ran into something that seems to me like an issue. Consider the following code:
```
BIGNUM* bn_in = BN_new();
BN_set_word(bn_in, 123);
BN_set_negative(bn_in, 1);
std::cout << "In: " << BN_bn2dec(bn_in) << "\n";

OSSL_PARAM_BLD* bld = OSSL_PARAM_BLD_new();
OSSL_PARAM_BLD_push_BN(bld, "test", bn_in);

OSSL_PARAM* params = OSSL_PARAM_BLD_to_param(bld);
OSSL_PARAM* param  = OSSL_PARAM_locate(params, "test");

BIGNUM* bn_out = NULL;
OSSL_PARAM_get_BN(param, &bn_out);
std::cout << "Out: " << BN_bn2dec(bn_out) << "\n";
```

This constructs a `OSSL_PARAM_BLD` with a negative bignum, creates the `OSSL_PARAM*` from that, retrieves the parameter and gets the bignum. I would expect both in and out to show "-123".

However, the output I get is:
```
In: -123
Out: 123
```
So the negative flag seems to be lost somewhere.

This was only during some testing; we don't actually need negative bignums in parameters at this moment. So this does not have a high priority.

Is this indeed an issue or unclear documentation, or am I missing something?

Thanks in advance!

Best regards,
SpoofedEx

**Comment by levitte** [comment]
Correct observation.  I thought this was documented, but now see that we don't.  That should be fixed.

This goes back to `BN_bn2bin`, which is documented like this:

> **BN_bn2bin()** converts the absolute value of **a** into big-endian form and
> stores it at **to**. **to** must point to **BN_num_bytes(a)** bytes of memory.

**Comment by paulidale** [comment]
The parameters builder does not support negative big numbers.
Likewise our parameter mechanism doesn't support them.

We need to fix the documentation.

**Comment by mattcaswell** [comment]
> The parameters builder does not support negative big numbers.

Shouldn't it fail if passed a negative number?

**Comment by levitte** [comment]
Proposed fix in #17074
